Add TLS and basic auth policy (#1611)
prometheus/docs: content/docs/operating/security.md

release date with you, acknowledging your effort and mentioning you by name
if you want.

Prometheus is maintained by volunteers, not by a company. Therefore, fixing on

> security issues is done on a best-effort basis. We thrive to fix security issues

Reviewer: s/thrive/strive

> in less than 7 days.

Reviewer 2: How about:

```
We strive to release security fixes within 7 days for (list of projects).
```

> in less than 7 days.

Reviewer 3: That's a question for @beorn7 really as the maintainer.

The reason that blackbox is in there is that Ben asked for it in that dev summit, and I said sure.

> in less than 7 days.

Reviewer 4: I think the wording is sufficiently vague that I'd be happy to do this kind of commitment for any repository I'm maintaining. (I mean, what would it mean to not do that? "I won't even _try_ to fix a security issue in a week time.")

> in less than 7 days.

Author: I take that as a yes, the reason behind explicitly covering it is that it is one that is exposed to more than just  the prometheus server.

We can find a consensus to cover more or all our repos in the coming dev summit.

release date with you, acknowledging your effort and mentioning you by name
if you want.

> Prometheus is maintained by volunteers, not by a company. Therefore, fixing on

Reviewer 5: `fixing on security issues` should perhaps just be `fixing security issues`.

Author: fixed

security issues is done on a best-effort basis. We strive to release security

> fixes within 7 days for: Prometheus, Alertmanager, Node_exporter,

Reviewer 6: I'd either say "node_exporter" or "Node Exporter" (probably the latter in a normal English sentence), but mixing snake_case with CamelCase seems odd (same for Blackbox Exporter).

Author: ack

> fixes within 7 days for: prometheus, alertmanager, node_exporter,

Reviewer 5: These project names are capitalized in the `Authentication, Authorization, and Encryption` section, consistency would be nice.

Author: done

> blackbox_exporter and pushgateway.

Author: Added comma before and

## Authentication, Authorization, and Encryption

In the future, server-side TLS support will be rolled out to the different
Prometheus projects. Those projects include Prometheus, Alertmanager,
Pushgateway and the official exporters.

> TLS client-side authentication with certificates will also be supported.

Reviewer 3: This seems misleading, we already support that for clients. "Authentication of clients by TLS client certs" would be clearer.

Author: fixed wording

The Go projects will share the same TLS library, which will be based on the
Go vanilla TLS library. We keep [Go default TLS
parameters](https://golang.org/pkg/crypto/tls/#Config), with one exception: we

> support only TLSv1.2 and higher.

Reviewer 3: This still needs a justification of some form, this appears completely arbitrary currently which will make it more challenging to deal with requests to change it good/bad reasons.

Author: I have added links.

Reviewer 3: That's a good start, we need something to cover why we're not changing anything else - otherwise we could have fun when any of those 4 organisations propose e.g. going to TLS 1.3 only or disabling specific ciphers. For example the current standing policy is to stick to the Go defaults, so that the decision is completely out of our hands and we thus can keep it simple and not worry about any of this.

What I'm looking for here (and have been asking for in relation to TLS&friends for many years) is essentially a plan that doesn't boil down to making it up as we go along. There's various hurdles we will inevitably have to deal with as things naturally evolve over time, and I don't want us to have a half-baked approach to TLS that considers the feature done and dusted merely because some code has been written. The code is the easy bit, and a justification that only covers why we have these particular settings right now is not enough. Users requesting changes to the hard coding, vulnerabilities, increases in computing power, geopolitics, changes in dependencies, and new standards are all going to happen.
The policy needs to be able to help answer questions like when we would disable 1.2, reenable 1.0 (and yes, it has happened in the past that an older version was suddenly more secure than a newer version), switch to/from whatever the cipher flavour of the month is, and what happens when Go changes their defaults. So things to be considered might include how much breakage of existing users is okay, is it okay to create a situation where the latest releases of official Prometheus components don't work with each other, how far back are we willing to have such breakage, how perfect should our ssllabs rating be, how good the actual security is for typical users, how often we need to reevaluate/update this configuration, how much effort we're realistically able/willing to put into this, and how we try to balance all these concerns.

Now all of that doesn't belong in this document as this document is more a what than why, but if you want to have a consensus to change our policy to one that allows for TLS 1.0/1.1 to be disabled then this needs to exist in some form.

Author: "We strive to keep TLS
compatibility with official client libraries
runtimes that were released less than two years ago."

Reviewer 3: Doing some research Go was the last to release TLS 1.3 support in February 2019, so 2 years means we would drop 1.2 support in February 2021.

TLS will be added to Java exporters in the future.

If you have special TLS needs, like a different cipher suite or older TLS
version, the TLS support enables you to build a secure tunnel between the
servers and reverse proxies with special settings.

HTTP Basic Authentication will also be supported. In such a case, usernames
and passwords are provided in the same configuration file as the TLS
configuration. Basic Authentication can be used without TLS, but it will then
expose usernames and passwords over the network.

Passwords are hashed with the [bcrypt](https://en.wikipedia.org/wiki/Bcrypt)
algorithm, and it is your responsibility to pick the number of rounds that
matches your security standards. More rounds make brute-force more complicated
at the cost of more CPU power and more time to authenticate the requests.

Various Prometheus components support client-side authentication and
encryption. If TLS client support is offered, there is often also an option
called `insecure_skip_verify` which skips SSL verification.

> # API calls

Reviewer 3: "API security" maybe?

Author: good idea

In the future, server-side TLS support will be rolled out to the different
Prometheus projects. Those projects include Prometheus, Alertmanager,
Pushgateway and the official exporters.

Authentication of clients by TLS client certs will also be supported.

The Go projects will share the same TLS library, which will be based on the
Go vanilla TLS library. We keep [Go default TLS
parameters](https://golang.org/pkg/crypto/tls/#Config), with one exception: we
support only TLSv1.2 and higher. This decision is based upon current industry
standards (notes from [Google][tlsgoogle], [Apple][tlsapple],
[Mozilla][tlsmozilla], and [Microsoft][tlsmicrosoft]).

TLS will be added to Java exporters in the future.

If you have special TLS needs, like a different cipher suite or older TLS
version, the TLS support enables you to build a secure tunnel between the
servers and reverse proxies with special settings.

HTTP Basic Authentication will also be supported. Basic Authentication can be
used without TLS, but it will then expose usernames and passwords in cleartext
over the network.

On the server side, basic authentication passwords are hashed on disk with the

> [bcrypt](https://en.wikipedia.org/wiki/Bcrypt) algorithm on disks, and it is

Reviewer 7: "are hashed on disk with ... algorithm~~ on disks~~,"

Author: Thx for noticing.

> On the server side, basic authentication passwords are hashed on disk with the
> [bcrypt](https://en.wikipedia.org/wiki/Bcrypt) algorithm on disks, and it is

Reviewer 2: ```suggestion
On the server side, basic authentication passwords are stored as hashes with the
[bcrypt](https://en.wikipedia.org/wiki/Bcrypt) algorithm. It is
```

Author: thx!

## Authentication, Authorization, and Encryption

In the future, server-side TLS support will be rolled out to the different
Prometheus projects. Those projects include Prometheus, Alertmanager,
Pushgateway and the official exporters.

Authentication of clients by TLS client certs will also be supported.

The Go projects will share the same TLS library, which will be based on the
Go vanilla TLS library. We keep [Go default TLS
parameters](https://golang.org/pkg/crypto/tls/#Config), with one exception: we
support only TLSv1.2 and higher. This decision is based upon current industry
standards (notes from [Google][tlsgoogle], [Apple][tlsapple],
[Mozilla][tlsmozilla], and [Microsoft][tlsmicrosoft]). We strive to keep TLS

> compatibility with official [client libraries](../instrumenting/clientlibs)

Reviewer 3: client library runtimes

This still isn't covering all the other settings

## Authentication, Authorization, and Encryption

In the future, server-side TLS support will be rolled out to the different
Prometheus projects. Those projects include Prometheus, Alertmanager,
Pushgateway and the official exporters.

Authentication of clients by TLS client certs will also be supported.

The Go projects will share the same TLS library, which will be based on the
Go vanilla [crypto/tls](https://golang.org/pkg/crypto/tls) library.
We default to TLS 1.2 as minimum version. Our policy regarding this is based on
[Qualys SSL Labs](https://www.ssllabs.com/) recommendations, where we strive to
achieve a grade 'A' with a default configuration and correctly provided
certificates, while sticking as closely as possible to the upstream Go defaults.
Achieving that grade provides a balance between perfect security and usability.

TLS will be added to Java exporters in the future.

If you have special TLS needs, like a different cipher suite or older TLS
version, you can tune the minimum TLS version and the ciphers, as long as the

> cipher are not [marked as insecure](https://golang.org/pkg/crypto/tls/#InsecureCipherSuites)

Reviewer 6: ```suggestion
cipher is not [marked as insecure](https://golang.org/pkg/crypto/tls/#InsecureCipherSuites)
```

Author: ack

code outside of our control or due to functionality that happens to expose
wherever it is stored.

> The web security configuration file, with TLS and HTTP Basic Authentication

Author: I have removed that

It is recommended to monitor all components for failure, and to have them
automatically restart on failure.

> ## Server-side TLS and authentication

Reviewer 3: This doesn't mention about basic-auth, and we should probably put in a warning about using that over the clear at least.

Author: Added basic auth

In the future, server-side TLS support will be rolled out to the different
projects under the Prometheus organization. Those projects include Prometheus,
Alertmanager, Pushgateway and the official exporters. The projects will share

> the same TLS library, which will be based on the Golang vanilla TLS library. We

Author: I have now tried to address non-go exporters

keep [Golang's default TLS
parameters](https://golang.org/pkg/crypto/tls/#Config), with one exception: we

> explicitly disable TLSv1.0 and TLSv1.1.

Reviewer 8: I feel this is worth re-discussing. At the time we didn't have TLSv1.3, I think we should default to that. Bonus points we get from that is that there are no configurable cipher suites with 1.3 yet anyways.

Reviewer 3: That'd in addition to already breaking client_java and client_python compared to Go defaults, would also break client_golang (and possibly also client_ruby), plus I imagine a non-trivial amount of users.

Reviewer 2: I would word this as "we support TLSv1.2 and higher".

Author: There is a difference between default to that and that only. I think that TLS1.3 is not spread enough to disable 1.2 entirely. Imagine that you run RHEL and you can't use curl to debug.

Reviewer 8: "TLSv1.2 and higher" I'm happy with as well. Needs to be configurable then though.

Author: rephrased to use 1.2 and higher. The configuration bit I have not added as I see no consensus.

I think that we can discuss the configuration part at the next dev summit to reach a consensus.

Reviewer 3: If you run RHEL when you need TLS 1.0 :)

Reviewer 3: Yeah, configuration is something that is to be avoided - that's what a reverse proxy is for as it stands.

This doc still doesn't explain why we stick with defaults, or why we make an exception for 1.2. Defaults was for general sanity/maintainability, but a general basis for cherry-picking versions (and why only versions, not ciphers or the other 10+ settings) needs to be elucidated.

> There is no plans at the moment to extend those settings. If you need older TLS

Reviewer 3: The first sentence is a bit wishy washy, I'd remove it.

Author: Rephrased that.

> There is no plans at the moment to extend those settings. If you need older TLS

Author: Rephrased it.

> versions or different cipher suite, we recommend you to setup a reverse proxy in

Reviewer 3: a different

Author: Rephrased that.

front of your services, which will handle those settings.

> Changes regarding those matters need to reach consensus within the

Author: I have removed it. It was to address your concerns about the PR's in common.

It is recommended to monitor all components for failure, and to have them
automatically restart on failure.

> ## Server-side TLS and authentication

Author: changed line 119

In the future, server-side TLS support will be rolled out to the different
projects under the Prometheus organization. Those projects include Prometheus,
Alertmanager, Pushgateway and the official exporters.

The golang projects will share the same TLS library, which will be based on the

> Golang vanilla TLS library. We keep [Golang's default TLS

Author: renamed

parameters](https://golang.org/pkg/crypto/tls/#Config), with one exception: we
support only TLSv1.2 and higher.

> Research will be made later on the non-golang exporters to see how to best

Reviewer 3: Java may end up being weird (e.g. how do you only support 1.2+ and the various client auth options in a JVM that doesn't support it? Will it even compile?), so I'd phrase it more as "TLS will be added to Java exporters in the future".

We've no Python or Ruby exporters, so we can limit this to Java.

Author: Rephrased

address server-side TLS, but you can expect a policy close to this one.

> The TLS support on the projects should enable you to build a secure tunnel

Reviewer 3: I'd predicate this on the special needs first, as currently written it makes it sounds like the primary purpose of this is the tunnel rather than this being primarily intended to be used standalone.

Author: addressed

between the servers and your reverse proxies if you have more specialized TLS
needs, like a cipher suite or older TLS version.

> Server-side Basic Authentication is also supported. In such a case, usernames

Reviewer 3: HTTP Basic Authentication

Author: fixed

and passwords are provided in the same configuration file as the TLS
configuration. Basic Authentication can be used without TLS, but it will then
expose usernames and passwords over the network.

Passwords are hashed with the [bcrypt](https://en.wikipedia.org/wiki/Bcrypt)
algorithm, and it is your responsibility to pick the number of rounds that
matches your security standards. More rounds make brute-force more complicated
at the cost of more CPU power and more time to authenticate the requests.

The web security configuration file, with users and path to certificates, can be
changed on the fly, with the exception of enabling/disabling TLS entirely. It is

> your responsibility to protect that file from unwanted reads/writes.

Reviewer 3: Hmm, this final bit would be better covered up in the Secrets section I think.

Author: moved

It is recommended to monitor all components for failure, and to have them
automatically restart on failure.

## Server-side TLS and authentication

In the future, server-side TLS support will be rolled out to the different
projects under the Prometheus organization. Those projects include Prometheus,
Alertmanager, Pushgateway and the official exporters.

The Go projects will share the same TLS library, which will be based on the
Go vanilla TLS library. We keep [Go default TLS
parameters](https://golang.org/pkg/crypto/tls/#Config), with one exception: we
support only TLSv1.2 and higher.

TLS will be added to Java exporters in the future.

If you have special TLS needs, like a different cipher suite or older TLS

> version, the TLS support on the projects should enable you to build a secure

Reviewer 3: This reads oddly. Maybe:

the TLS support enables you to

Author: done

tunnel between the servers and reverse proxies with special settings.

HTTP Basic Authentication is also supported. In such a case, usernames

> and passwords are provided in the same configuration file as the TLS

Reviewer 3: Hmm, this is the first mention of a configuration file which makes it seem as though we forgot that - though I don't think we need this level of detail in this doc (that's more for the usage docs).

Author: that config file has security implications -- if people give access to anyone it is possible to change certs, disable client side auth, add users..

Reviewer 3: What goes in what config files is too detailed to go here I think, plus other config files also have security implications depending on your setup.

Fields containing secrets in configuration files (marked explicitly as such in
the documentation) will not be exposed in logs or via the HTTP API. Secrets
should not be placed in other configuration fields, as it is common for
components to expose their configuration over their HTTP endpoint. It is the

> responsibility of the user to protect files on disks from unwanted reads and

Reviewer 3: on disk

code outside of our control or due to functionality that happens to expose
wherever it is stored.

Special care must be taken to prevent unauthorized reads or writes to

> configuration files that contain secrets.

Reviewer 3: That's not really Prometheus specific, nor secret specific. I'd add something two paragraphs up that protecting the files on disk is the user's responsibility.

Author: ack
